app/main/views/direct_award.py

```diff
@@ -348,3 +343,71 @@ def record_project_download(project_external_id):
     db.session.commit()
 
     return single_result_response("project", project), 200
+
+
+@main.route('/direct-award/projects/<int:project_external_id>/services/<string:service_id>/award', methods=['POST'])
+def create_award_process_outcome(project_external_id, service_id):
+    updater_json = validate_and_return_updater_request()
+
+    # Acquire share-lock on project and active search rows - what we're trying to assert here is that the "locked" field
+    # of the project and the "active" field of the search are set to appropriate values at the time we create this
+    # process outcome
+    project = db.session.query(DirectAwardProject).options(
+        # this has to be performed as an inner join for the row-lock to work, just means we have to perform an extra
+        # database hit in the no-rows case to disambiguate the error message
+        db.joinedload(DirectAwardProject.active_search, innerjoin=True)
+    ).filter(
+        DirectAwardProject.external_id == project_external_id,
+    ).with_for_update(read=True).first()
+
+    if project is None:
+        if db.session.query(DirectAwardProject.id).filter_by(external_id=project_external_id).first():
+            # by deduction...
+            abort(404, f"Project {project_external_id} doesn't have an active search")
+        else:
+            abort(404, f"Project {project_external_id} not found")
+
+    if not project.locked_at:
+        abort(400, f"Project {project_external_id} has not been locked")
+
+    if project.process_outcome is not None:
+        abort(410, "Project {} already has a completed process outcome: {}".format(
+            project_external_id,
+            project.process_outcome.external_id,
+        ))
+
+    archived_service = project.active_search.archived_services.options(db.load_only("id")).filter_by(
+        service_id=service_id
+    ).first()
+
+    if not archived_service:
+        abort(404, f"Service {service_id} is not in project {project_external_id}'s active search search results")
+
+    # TODO try-loop to assign ids, handling (unlikely) external_id collisions using sub-transaction so we don't lose
```

Comment:
Is this a problem we have elsewhere in the codebase? Do we need a TODO if we can make a tech debt ticket? If it's urgent enough for a `TODO` shouldn't we be doing it now?

Reply:
Pass. It's very unlikely until we start to have *many* rows in the table. Don't know whether people want me to bother or not, but wanted to properly record how the design intends for it to be done when/if someone comes along to do it.